applications/StructuralMechanicsApplication/custom_processes/apply_multi_point_constraints_process.h

```diff
@@ -32,134 +29,206 @@
+        } else if (dim == 3) {
+            ApplyConstraints<3>(r_master_model_part, r_slave_model_part);
         }
     }
 
     // Functions which use two variable components
     template <int TDim>
-    void ApplyConstraints(ModelPart &master_model_part, ModelPart &slave_model_part)
+    void ApplyConstraints(
+        ModelPart &rMasterModelPart,
+        ModelPart &rSlaveModelPart
+        )
     {
-        BinBasedFastPointLocator<TDim> *p_point_locator = new BinBasedFastPointLocator<TDim>(master_model_part);
-        int numVars = m_parameters["variable_names"].size();
+        BinBasedFastPointLocator<TDim> *p_point_locator = new BinBasedFastPointLocator<TDim>(rMasterModelPart);
+        const SizeType num_vars = mParameters["variable_names"].size();
         // iterating over slave nodes to find the corresponding masters
-        const int n_slave_nodes = slave_model_part.Nodes().size();
+        const SizeType n_slave_nodes = rSlaveModelPart.Nodes().size();
         array_1d<double, TDim + 1> N; // This is only for triangular meshes
         const int max_results = 100;
         typename BinBasedFastPointLocator<TDim>::ResultContainerType results(max_results);
 
-        for (int i = 0; i < n_slave_nodes; i++)
-        {
-            ModelPart::NodesContainerType::iterator iparticle = slave_model_part.NodesBegin() + i;
-            Node<3>::Pointer p_slave_node = *(iparticle.base());
+        for (IndexType i = 0; i < n_slave_nodes; i++) {
+            ModelPart::NodesContainerType::iterator iparticle = rSlaveModelPart.NodesBegin() + i;
+            NodeType::Pointer p_slave_node = *(iparticle.base());
             typename BinBasedFastPointLocator<TDim>::ResultIteratorType result_begin = results.begin();
-            Element::Pointer pMasterElement;
-            bool is_found = false;
-            is_found = p_point_locator->FindPointOnMesh(p_slave_node->Coordinates(), N, pMasterElement, result_begin, max_results);
-            if (is_found == true)
-            {
-                for (int i = 0; i < numVars; i++)
-                {
-                    std::string varName = m_parameters["variable_names"][i].GetString();
-                    Geometry<Node<3>> &geom = pMasterElement->GetGeometry();
-                    for (unsigned int i = 0; i < geom.size(); i++)
-                    {
-                        if (KratosComponents<Variable<double>>::Has(varName))
-                        { //case of double variable
-                            VariableType rVar = KratosComponents<Variable<double>>::Get(m_parameters["variable_names"][i].GetString());
+            Element::Pointer p_master_element;
+            const bool is_found = p_point_locator->FindPointOnMesh(p_slave_node->Coordinates(), N, p_master_element, result_begin, max_results);
+            if (is_found == true) {
+                for (IndexType i = 0; i < num_vars; i++) {
+                    std::string varName = mParameters["variable_names"][i].GetString();
```

Comment:
this is a string access within a node loop?? to parameters??

absolutely to be avoided...

Reply:
Not my original code...

Reply:
@RiccardoRossi this means one has to loop over all the nodes on slave modelpart for each variable. instead I choose to loop over the nodes once and check using this string access. 
Do you suggest the other way is efficient ?

Reply:
Hi @adityaghantasala string manipulation/comparison is extremely slow and should be avoided as much as possible.

also the "Parameters" objects and the "KratosComponents" are designed more to catch errors than for fast performance. 

Long story short, i would definitely invert the loops (that is, do the variable loop as outer).

Just for curiosity time it. I am very convinced you'll see a 1-2 order of magnitude speedup.